File: examples/python/monitor_tasks.py
"""
NavaiaForge SDK — Monitor Tasks and Cost

Demonstrates listing tasks, checking statuses, and querying
observability metrics for a workforce.

Run:
    export NAVAIA_FORGE_BASE_URL="http://localhost:8000"
    export NAVAIA_FORGE_API_KEY="nf_..."
    export NAVAIA_FORGE_WORKFORCE_ID="wf_..."
    python examples/python/monitor_tasks.py
"""

from __future__ import annotations

import os
import sys

from navaia_forge import NavaiaForgeClient

_STATUS_ICONS: dict[str, str] = {
    "done": "[OK]",
    "in_progress": "[..]",
    "pending": "[--]",
    "failed": "[!!]",
    "rejected": "[XX]",
    "blocked": "[##]",
    "waiting": "[~~]",
}


def main() -> None:
    workforce_id = os.environ.get("NAVAIA_FORGE_WORKFORCE_ID")
    if not workforce_id:
        sys.stderr.write("Set NAVAIA_FORGE_WORKFORCE_ID before running.\n")
        sys.exit(1)

    client = NavaiaForgeClient(
        base_url=os.environ.get("NAVAIA_FORGE_BASE_URL", "http://localhost:8000"),
        api_key=os.environ.get("NAVAIA_FORGE_API_KEY", ""),
    )

    # ── 1. List all tasks ──────────────────────────────────────

    all_tasks = client.tasks.list(workforce_id)
    print(f"Total tasks: {len(all_tasks)}\n")

    for task in all_tasks:
        icon = _STATUS_ICONS.get(task.status, "[??]")
        print(f"  {icon} {task.title}")
        print(f"       ID: {task.id}")
        print(f"       Status: {task.status}")
        print(f"       Created: {task.created_at}")
        if task.completed_at:
            print(f"       Completed: {task.completed_at}")
        print()

    # ── 2. Filter by status ───────────────────────────────────

    in_progress = client.tasks.list(workforce_id, status="in_progress")
    failed = client.tasks.list(workforce_id, status="failed")
    print(f"In progress: {len(in_progress)}")
    print(f"Failed: {len(failed)}")

    # ── 3. Inspect a specific task ─────────────────────────────

    if all_tasks:
        latest = all_tasks[0]
        detail = client.tasks.get(latest.id)
        print("\nLatest task detail:")
        print(f"  Title:       {detail.title}")
        print(f"  Description: {detail.description}")
        print(f"  Status:      {detail.status}")
        print(f"  Agent:       {detail.agent_id or 'unassigned'}")
        print(f"  Priority:    {detail.priority}")
        if detail.result:
            print(f"  Result:      {detail.result[:200]}...")
        if detail.error:
            print(f"  Error:       {detail.error}")

    # ── 4. Observability — metrics summary (raw dict payload) ──

    metrics = client.observability.summary(workforce_id)
    print("\nMetrics Summary:")
    print(f"  Total tasks:       {metrics.get('total_tasks', 0)}")
    print(f"  Completed:         {metrics.get('completed_tasks', 0)}")
    print(f"  Failed:            {metrics.get('failed_tasks', 0)}")
    print(f"  Active agents:     {metrics.get('active_agents', 0)}")
    print(f"  Tokens today:      {metrics.get('total_tokens_today', 0):,}")
    print(f"  Cost today:        ${metrics.get('cost_today', 0):.4f}")

    # ── 5. Cost breakdown over the last 7 days ────────────────

    cost = client.observability.cost(workforce_id, days=7)
    print(f"\nCost (last {cost.period_days} days):")
    print(f"  Total tokens:    {cost.total_tokens:,}")
    print(f"  Weighted tokens: {cost.total_weighted_tokens:,}")
    print(f"  Total cost:      ${cost.total_cost_usd:.4f}")

    if cost.by_model:
        print("\n  By model:")
        for row in cost.by_model:
            print(
                f"    {row.model:<20} "
                f"tokens={row.total_tokens:>10,}  "
                f"cost=${row.cost_usd:.4f}"
            )

    # ── 6. Retry failed tasks ─────────────────────────────────

    if failed:
        print(f"\nRetrying {len(failed)} failed task(s)...")
        for task in failed:
            retried = client.tasks.retry(task.id)
            print(f"  Retried: {retried.title} -> {retried.status}")


if __name__ == "__main__":
    main()

File: examples/javascript/monitor-tasks.ts
/**
 * NavaiaForge SDK — Monitor Tasks and Token Usage
 *
 * Demonstrates listing tasks, checking statuses, and querying
 * observability metrics for a workforce.
 */

import { NavaiaForge } from "@navaia/forge";

const nf = new NavaiaForge({
  apiKey: "nf_your_api_key",
  baseUrl: "https://api.navaia.com",
});

const WORKFORCE_ID = "wf_your_workforce_id";

// ── 1. List all tasks ──────────────────────────────────────

const allTasks = await nf.tasks.list(WORKFORCE_ID);
console.log(`Total tasks: ${allTasks.length}\n`);

const statusIcons: Record<string, string> = {
  done: "[OK]",
  in_progress: "[..]",
  pending: "[--]",
  failed: "[!!]",
  rejected: "[XX]",
};

for (const task of allTasks) {
  const icon = statusIcons[task.status] ?? "[??]";
  console.log(`  ${icon} ${task.title}`);
  console.log(`       ID: ${task.id}`);
  console.log(`       Status: ${task.status}`);
  console.log(`       Created: ${task.created_at}`);
  if (task.completed_at) {
    console.log(`       Completed: ${task.completed_at}`);
  }
  console.log();
}

// ── 2. Filter by status ───────────────────────────────────

const inProgress = await nf.tasks.list(WORKFORCE_ID, "in_progress");
const failed = await nf.tasks.list(WORKFORCE_ID, "failed");
console.log(`In progress: ${inProgress.length}`);
console.log(`Failed: ${failed.length}`);

// ── 3. Inspect a specific task ─────────────────────────────

if (allTasks.length > 0) {
  const latest = allTasks[0]!;
  const detail = await nf.tasks.get(latest.id);

  console.log(`\nLatest task detail:`);
  console.log(`  Title:       ${detail.title}`);
  console.log(`  Description: ${detail.description}`);
  console.log(`  Status:      ${detail.status}`);
  console.log(`  Agent:       ${detail.agent_id ?? "unassigned"}`);
  console.log(`  Priority:    ${detail.priority}`);
  if (detail.result) {
    console.log(`  Result:      ${detail.result.slice(0, 200)}...`);
  }
  if (detail.error) {
    console.log(`  Error:       ${detail.error}`);
  }
}

// ── 4. Observability — metrics summary ─────────────────────

const metrics = await nf.observability.summary(WORKFORCE_ID);
console.log(`\nMetrics Summary:`);
console.log(`  Total tasks:       ${metrics.total_tasks}`);
console.log(`  Completed:         ${metrics.completed_tasks}`);
console.log(`  Failed:            ${metrics.failed_tasks}`);
console.log(`  Active agents:     ${metrics.active_agents}`);
console.log(`  Tokens today:      ${metrics.total_tokens_today.toLocaleString()}`);
console.log(`  Cost today:        $${metrics.cost_today.toFixed(4)}`);

// ── 5. Token usage breakdown ──────────────────────────────

const tokenHistory = await nf.observability.tokenUsage(WORKFORCE_ID, 7);
console.log(`\nToken usage (last 7 days): ${tokenHistory.length} records`);

const totalInput = tokenHistory.reduce((sum, r) => sum + r.input_tokens, 0);
const totalOutput = tokenHistory.reduce((sum, r) => sum + r.output_tokens, 0);
const totalCost = tokenHistory.reduce((sum, r) => sum + r.cost_weighted, 0);

console.log(`  Input tokens:  ${totalInput.toLocaleString()}`);
console.log(`  Output tokens: ${totalOutput.toLocaleString()}`);
console.log(`  Total cost:    $${totalCost.toFixed(4)}`);

// ── 6. Retry failed tasks ─────────────────────────────────

if (failed.length > 0) {
  console.log(`\nRetrying ${failed.length} failed task(s)...`);
  for (const task of failed) {
    const retried = await nf.tasks.retry(task.id);
    console.log(`  Retried: ${retried.title} -> ${retried.status}`);
  }
}
